File: day2.py
#!/usr/bin/python3
from shared import read_from_file
import re

passwords = read_from_file('inputs/day2.txt')

def is_valid_password_count(passwords):
  valid_count = 0
  for password in passwords:
    first_num, second_num, character, pwd = list(filter(None,re.split('-|:|\\s', password)))
    first_num, second_num = [int(x) for x in [first_num, second_num]]
    if pwd.count(character) >= first_num and pwd.count(character) <= second_num:
      valid_count += 1
  return valid_count

def is_valid_password_char(passwords):
  valid_count = 0
  for password in passwords:
    first_num, second_num, character, pwd = list(filter(None,re.split('-|:|\\s', password)))
    first_num, second_num = [int(x) for x in [first_num, second_num]]
    if pwd[first_num - 1] == pwd[second_num  - 1]: 
      continue
    if pwd[first_num - 1] == character or pwd[second_num - 1] == character:
      valid_count += 1
  return valid_count

print("Part 1", is_valid_password_count(passwords))
print("Part 2", is_valid_password_char(passwords))

File: node/day2.js
const tools = require('./tools')

const passwords = tools.readFromFile('../inputs/day2.txt')

function validPasswordCount (passwords) {
  validCount = 0
  passwords.forEach((p) => {
    let [minNum, maxNum, char, pwd] = p.split(/-|:|\s/g).filter(p => p)
    // if (new RegExp(char, 'g') || []).length
    let re = new RegExp(char, 'g')
    count_chars = (pwd.match(re) || []).length
    if (count_chars >= minNum && count_chars <= maxNum) {
      validCount++
    }
  })
  return validCount
}

function validPasswordChar (passwords) {
  validCount = 0
  passwords.forEach((p) => {
    let [minNum, maxNum, char, pwd] = p.split(/-|:|\s/g).filter(p => p)
    if ((pwd[minNum - 1] === char || pwd[maxNum - 1] === char) && pwd[minNum - 1] !== pwd[maxNum - 1]) {
      validCount++
    }
  })
  return validCount
}

console.log(`Part 1: ${validPasswordCount(passwords)}`)
console.log(`Part 2: ${validPasswordChar(passwords)}`)